Getweight

PieterTack opened the pull request.

PieterTack committed:
polycap-photon: change energy dependency for photon_new to photon_launch
Issue #22

PieterTack commented:
deze build faalt. Segmentation fault in `tests/capil.c`, maar geen idee waarom of waar. Heb valgrind geprobeerd, maar raak er niet wijzer van: `valgrind --leak-check=full --track-origins=yes --show-reachable=yes ./src/polycap ../example/dub_foc.inp`
Zou eigenlijk valgrind op de make check moeten doen, maar geen idee hoe.

In python stuk heb ik ook nogal raar gedaan vrees ik... krijg er heel pak warnings en errors, wss omdat ik syntax ergens fout heb nagelaten...

tschoonj commented:
Ge hebt in elk geval uw branch verkeerd genomen: uw commits uit de vorige pull request zitten hier ook :-)

Doe het volgende (terwijl ge op de `getweight` branch zit):

```
git fetch
git rebase origin/master
# als er geen problemen zijn doe dan:
git push -f
```

In het vervolg moet ge altijd een nieuwe branch maken tov `master`:

```
git checkout master
git pull
git checkout -b new-branch-name # maakt een nieuwe branch en gebruikt hem direct
```

tschoonj commented:
Ok thanks, zal even wat debuggen.

tschoonj committed:
Fix tests after API changes

PieterTack commented:
Ah, de weights :) Of course... silly me.

tschoonj commented:
Bij mij heeft valgrind die makkelijk gevonden... bizar

tschoonj committed:
Fix unused variable warnings

PieterTack commented:
Maar wat wat je command line dan voor valgrind? Want ik had bij mij de indruk dat hij het gewoon op de normale code deed, maar daar het fout in de tests zelf zat gaf hij niks merkwaardig weer.

Dus maw: hoe run ik valgrind tijdens de make check?

tschoonj commented:
Zonet mijn commit gefikst voor nog wat unused variable warnings eruit te halen.

Je zal nu moeten `git pull --rebase` doen bij u.

Om valgrind te gebruiken moet je eerst zien dat alle test executables gebuild zijn. Dus:

```
make
make check
```

Daarna ga je in `tests` en doe je bijvoorbeeld:

```
valgrind --tool=memcheck ./capil
```

PieterTack commented:
ah, I see :) Thanks!

PieterTack committed:
polycap_photon_launch: add double **weights argument
program currently runs, but gives wrong output
maybe due to malloc in routine which is called multiple times without free?
additionally, python bindings are still messed up

PieterTack commented:
Kprobeer die `double **weights` in `polycap_photon_launch` te implementere. Beetje gevecht met die pointers to pointers, maar heb nu toch een 'werkende' versie. Output transmissie efficientie is echter fout (lager dan anders).
Kan dit zijn omdat ik in `polycap_photon_launch` zelf `malloc`, maar deze functie vervolgens meerdere malen wordt geroepen voor ik eigenlijk de kans heb om `weights` te free'en?

Kweet niet of ik, in het oog van de veranderingen die ook in `polycap-source.c` dienen gemaakt te worden niet meer fan ben van een normale `polycap_photon_get_weights` functie. Dat is dan natuurlijk wel een hoop veranderingen die we net gemaakt hebben die dan strikt genomen niet nodig waren, maar oke :P

tschoonj commented:
Ben je zeker dat deze commit verantwoordelijk is voor de foute transmissie efficiëntie of is dat al eerder gebeurd?

Ik denk dat je best de unit test aanpast zodat de bekomen transmissie efficiëntie vergeleken wordt met de verwachtte, zodat we direct weten als er iets misgaat.

PieterTack commented:
Hm, heb snel een testje gerund (basically snel mijn laatste aanpassingen zoveel mogelijk ongedaan gemaakt) en de efficientie blijft lager. Probleem heeft zich blijkbaar dus al eerder voorgedaan.
Khad inderdaad eerder al een unit test hiervoor moeten maken. Geen idee waarom ik dat destijds over het hoofd heb gezien.

Hoe kan ik in git terugrollen naar vorige versies? Kzou dat wat willen proberen om te zien of ik daadwerkelijk nog de oude resultaten terugkrijg, of dat er iets anders vreemd aan de hand is. Immers met de laatste veranderingen (`energies` naar `polycap_photon_launch` ipv `polycap_photon_new`) zie ik niet in hoe dat plots invloed kan hebben op de transmissie efficienties, behalve dan als er memory issues zijn in geslopen.
Kzou bijvoorbeeld enerzijds terug willen naar 28a0186 (versie waarin al die figuren van poster gemaakt zijn, toen ik in Grenoble zat nog) en dan naar ad9ba05 (versie voor ik nu met die `double **weights` ben beginnen spelen)

tschoonj commented:
Teruggaan naar een vroegere commit doe je door `git checkout checksum`

Met checksum bedoel ik de string met hexadecimale karakters die elke commit uniek maakt.

Probeer eens `git log` en je zal de history zien, met checksums.

Je kan ook eens [`git bisect`](https://git-scm.com/docs/git-bisect) bekijken, een git commando dat veel gebruikt wordt in situaties als deze (maar nog nooit door mij...)

PieterTack commented:
oke, thanks :) Zal dat eens proberen (die `checkout`, de bisect lijkt me voorlopig nog net iets te technisch, maar je weet nooit dat ik een avontuurlijke bui raak natuurlijk)
En sorry dat ik het vroeg, vlak nadat ik het postte raadpleegde ik de GoogleGods en die wisten me in exact dezelfde tijd hetzelfde te vertellen. (Does this mean you are secretly a GoogleGod, or have equal powers? Interesting...)

tschoonj commented:
Wel misschien heeft het feit dat ik al 10 jaar git/github gebruik er iets mee te maken 😄 

Die bisect lijkt eigenlijk wel cool... ik zou daar toch eens mee moeten zien te spelen.

PieterTack commented:
oke, het transmissie efficientie probleem doet zich dus voor ergens tussen 28a0186 en ad9ba05.
Zal dus in de 28a0186 versie een unit test ontwerpen (we weten immers dat die sort-of overeenkomt met experimentele waarden ;) ) en dan beginnen uitdokteren wat er mis gaat tegen ad9ba05 :)

PieterTack committed:
polycap-photon.c: add **weights to polycap_photon_launch
Including test to verify content of weights and photon->weight
Closes #22

PieterTack commented:
to be merged

tschoonj commented:
Hmmm nee, de weights worden niet gereturned in de python bindings.

Zal dat even fiksen.

tschoonj committed:
Python bindings: have Photon.launch return weights

tschoonj commented:
Ben geen grote van die `polycap_photon_launch` `int` return value: die kan ofwel 0 of -1 zijn, en in dat laatste geval wordt misschien de error gezet... dat is niet cool.

Volgens mij zal je hier toch (minstens, in het geval ik iets over het hoofd heb gezien), 3 verschillende return values moeten hebben.

PieterTack commented:
Mja, is gekomen doordat oorspronkelijk 0 of -1 returnde als waarde, maar met toevoegen van errors heb ik die ook als -1 gezet omdat dat een logischere 'error waarde' is.

Zal het wat aanpassen dat -1 returned bij error en anders 0 of 1, waarbij 0 dan falen en 1 slagen is ofzo. Sounds better?

tschoonj commented:
yep, klinkt veel beter 👍

PieterTack committed:
polycap-source: appropriate weights_temp free when simulating new photon

PieterTack committed:
polycap-photon: free array at end of polycap_photon_launch
build gives errors on several tests... no idea why

PieterTack committed:
Thank you valgrind

tschoonj commented:
😃

PieterTack commented on src/polycap-source.c:
```diff
@@ -579,7 +564,10 @@ polycap_transmission_efficiencies* polycap_source_get_transmission_efficiencies(
-			if(iesc == -1) polycap_photon_free(photon); //Free photon here as a new one will be simulated
+			if(iesc == -1) {
+				polycap_photon_free(photon); //Free photon here as a new one will be simulated
+//				free(weights_temp);
```
@tschoonj Als deze niet gecommentarieerd staat zorgt het voor veel memory leaks. Lijkt me echter dat die wel hier gefree moet worden (immers wordt geinitialiseerd in `polycap_photon_launch`)
Jij enig idee?

tschoonj replied:
Yep `weights_temp` moet zeker ge`free`'d worden aangezien deze gealloceerd wordt in `polycap_photon_launch` en op dit punt niet meer nodig is.

Waarom zou je het niet doen misschien? Krijg je een segmentatie fout ofzo?

PieterTack replied:
idd, segmentation fault.

Maar ben het op het spoor. Het is (vermoed ik) omdat als `polycap_photon_launch` -1 returned, het ook kan zijn voordat `weights_temp` geinitialiseerd is. Dus ben nu eerst die output van `polycap_photon_launch`  aan het aanpassen, en denk dat dan wel goed komt met deze free ook.

tschoonj commented on src/polycap-source.c:
```diff
@@ -566,7 +566,7 @@ polycap_transmission_efficiencies* polycap_source_get_transmission_efficiencies(
 			if(iesc == -1) {
 				polycap_photon_free(photon); //Free photon here as a new one will be simulated
-				free(weights_temp);
+//				free(weights_temp);
```
Beter dergelijke lijnen gewoon te verwijderen ipv ze uit te commentariëren.

PieterTack committed:
polycap_photon_launch: error-separate return values
Closes #22
Closes #21

PieterTack commented on python/polycap.pyx:
```diff
@@ -306,6 +306,8 @@ cdef class Photon:
         if rv == -1:
             return None
+        if rv == 0:
+            return None
```
Niet zeker of je dit de meest elegante manier vindt...
Dus nu is -1 een error, 0 dat photon polycapillair einde niet haalde (dus weights zeer klein)
Mss is het voor correctheid op zich wel beter om ook bij `rv == 0` toch de weights door te geven? Indien het geval dient de `python.py` tests wat aangepast te worden...

tschoonj replied:
Als de `error` gezet wordt, zal `set_exception` een exception throwen waardoor lijn 307 nooit bereikt wordt. 

Zijn de weights nuttig als het einde van het polycapillair toch niet bereikt is? Anders mag het gewoon `None` blijven van mij.

PieterTack replied:
>Zijn de weights nuttig als het einde van het polycapillair toch niet bereikt is? Anders mag het gewoon `None` blijven van mij.

Hm, goed punt. Idd niet nuttig. Ik was aan het redeneren dat de gewichten dan gewoon heel klein waren, maar in se wel bruikbaar. Maar inderdaad stopt de berekening op dat punt direct, dus het zou fout zijn om met die foton verder te rekenen of wat dan ook.

Laten we het dan zo laten :)
Mag dan gemerged worden voor mijn part.

tschoonj replied:
Kan je nog die check in `polycap.pyx` voor `rv == -1` eruit halen? Die lijn kan toch niet bereikt worden aangezien de error zal getransformeerd worden in een exception door `set_exception`.

Nieuwe commit niet nodig: gewoon de laatste amenderen en force pushen.

tschoonj commented:
Nu deze gemerged is: misschien eens kijken naar documentatie?

Ik had daar een paar weken terug al een beetje aan gewerkt maar niets gecommit: als je tijd en zin hebt om daar nu aan te werken open ik anders wel een PR met wat ik al gedaan heb...

PieterTack commented:
Yup, ging nu inderdaad documentatie doen :)
Natuurlijk eerst opzoeken wat er moet gebeuren, en wss nog wat beamtime data ertussen bekijken ook, maar wou zo dicht mogelijk tegen exrs al wat documentatie hebben staan :)

tschoonj commented:
ok cool. zal binnen dik uur PR openen
